packages/runtime/src/components/_internal/ImplWrapper/ImplWrapperMain.tsx

```diff
@@ -0,0 +1,171 @@
+      return () => stops.forEach(s => s());
+    }, [c.id, c.traits, executeTrait, stateManager]);
+
+    // reduce traitResults
+    const propsFromTraits: TraitResult<string, string>['props'] = useMemo(() => {
+      return Array.from(traitResults.values()).reduce(
+        (prevProps, result: TraitResult<string, string>) => {
+          if (!result.props) {
+            return prevProps;
+          }
+
+          return mergeWith(prevProps, result.props, (obj, src) => {
+            if (isArray(obj)) {
+              return obj.concat(src);
+            }
+          });
+        },
+        {} as TraitResult<string, string>['props']
+      );
+    }, [traitResults]);
+
+    // component properties
+    const [evaledComponentProperties, setEvaledComponentProperties] = useState(() => {
+      return merge(
+        stateManager.deepEval(c.properties, { fallbackWhenError: () => undefined }),
+        propsFromTraits
+      );
+    });
+    // eval component properties
+    useEffect(() => {
+      const { result, stop } = stateManager.deepEvalAndWatch(
+        c.properties,
+        ({ result: newResult }: any) => {
+          setEvaledComponentProperties({ ...newResult });
+        },
+        { fallbackWhenError: () => undefined }
+      );
+      // must keep this line, reason is the same as above
+      setEvaledComponentProperties({ ...result });
+
+      return stop;
+    }, [c.properties, stateManager]);
+
+    useEffect(() => {
+      propsFromTraits?.componentDidMount?.forEach(e => e());
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+
+    useDidUpdate(() => {
+      propsFromTraits?.componentDidUpdate?.forEach(e => e());
+    });
+
+    useDidUnmount(() => {
+      propsFromTraits?.componentDidUnmount?.forEach(e => e());
+    });
+
+    const mergedProps = useMemo(
+      () => ({ ...evaledComponentProperties, ...propsFromTraits }),
+      [evaledComponentProperties, propsFromTraits]
+    );
+
+    const unmount = traitResults.some(result => result.unmount);
+    const slotElements = useSlotElements(props);
+
+    const C = unmount ? null : (
+      <Impl
+        key={c.id}
+        {...props}
+        {...mergedProps}
+        slotsElements={slotElements}
+        mergeState={mergeState}
+        subscribeMethods={subscribeMethods}
+        elementRef={eleRef}
+        getElement={onRef}
+      />
+    );
+
+    const result = (
+      <React.Fragment key={c.id}>
+        {C}
+        {children}
+      </React.Fragment>
+    );
+
+    const element = useGridLayout(props, result, ref);
+
+    return element;
+  }
+);
+
+// This hook will only run unmount function when unmount, not every time when unmount function changes.
+const useDidUnmount = (fn: Function) => {
+  const fnRef = useRef(fn);
+  fnRef.current = fn;
+
+  useEffect(() => () => fnRef.current(), []);
+};
+
+// This hook will run every times when component update, except when first render.
+const useDidUpdate = (fn: Function) => {
```

Comment:
Maybe we should support the dependencies and cleanup in the update? Because the trait may execute some logic and clean up some side-effects only when the dependencies are updated instead of any properties being updated.

For example, I want to implement a trait to achieve this:
```js
useEffect(()=> {
  // when the source or callback is changed, create the new subscription.
  const subscription = subscribe(props.source, props.callback);

  return () => {
    // unsubscribe the old source
    subscription.unsubscribe();
  };
}, [props.source, props.callback]);

```

Reply:
OK. I have added return functions to  `componentDidUpdate` and `componentDidMount`